neverbleed.c

```diff
@@ -606,27 +852,81 @@ static int load_key_stub(struct expbuf_t *buf)
         strerror_r(errno, errbuf, sizeof(errbuf));
         goto Respond;
     }
-    if ((rsa = PEM_read_RSAPrivateKey(fp, NULL, NULL, NULL)) == NULL) {
+
+    if ((pkey = PEM_read_PrivateKey(fp, NULL, NULL, NULL)) == NULL) {
         snprintf(errbuf, sizeof(errbuf), "failed to parse the private key");
         goto Respond;
     }
-    key_index = daemon_set_rsa(rsa);
-    estr = BN_bn2hex(rsa->e);
-    nstr = BN_bn2hex(rsa->n);
+
+    switch (EVP_PKEY_base_id(pkey)) {
+    case EVP_PKEY_RSA: {
+        const BIGNUM *e, *n;
+
+        rsa = EVP_PKEY_get1_RSA(pkey);
+        type = NEVERBLEED_TYPE_RSA;
+        key_index = daemon_set_rsa(rsa);
+        RSA_get0_key(rsa, &n, &e, NULL);
+        estr = BN_bn2hex(e);
+        nstr = BN_bn2hex(n);
+        break;
+    }
+    case EVP_PKEY_EC: {
+#if OPENSSL_1_1_API
+        const EC_POINT *ec_pubkey;
+        EC_KEY *ec_key;
+
+        ec_key = EVP_PKEY_get0_EC_KEY(pkey);
+        type = NEVERBLEED_TYPE_ECDSA;
+        key_index = daemon_set_ecdsa(ec_key);
+        ec_group = EC_KEY_get0_group(ec_key);
+        ec_pubkey = EC_KEY_get0_public_key(ec_key);
+        ec_pubkeybn = BN_new();
+        if (!EC_POINT_point2bn(ec_group, ec_pubkey, POINT_CONVERSION_COMPRESSED, ec_pubkeybn, NULL)) {
+            type = NEVERBLEED_TYPE_NONE;
+            snprintf(errbuf, sizeof(errbuf), "failed to convert ECDSA public key to BIGNUM");
+            goto Respond;
+        }
+        ec_pubkeystr = BN_bn2hex(ec_pubkeybn);
+        break;
+#else
+        snprintf(errbuf, sizeof(errbuf), "ECDSA support requires OpenSSL >= 1.1.0");
+        goto Respond;
+#endif
+    }
+    default:
+        snprintf(errbuf, sizeof(errbuf), "unsupported private key: %d", EVP_PKEY_base_id(pkey));
+        goto Respond;
+    }
 
 Respond:
     expbuf_dispose(buf);
-    expbuf_push_num(buf, rsa != NULL);
+    expbuf_push_num(buf, type != NEVERBLEED_TYPE_NONE);
+    expbuf_push_num(buf, type);
```

Comment:
Would the code become more concise if `NEVERBLEED_TYPE_NONE` could be used as an indication of an error, instead of sending a boolean value indicating success _and_ the type value?
(and if we are to use the _none_ type as error indication it might make more sense to rename it the _error_ type)